src/figure_finder.py

```diff
@@ -0,0 +1,185 @@
+        ts_image_bboxes = message_filters.TimeSynchronizer([image_sub, bboxes_sub], 10)
+        ts_image_bboxes.registerCallback(self.detection_callback)
+
+        self.global_pos_sub = rospy.Subscriber(
+            "mavros/global_position/global", NavSatFix, self.global_pos_callback
+        )
+        self.rel_alt_sub = rospy.Subscriber(
+            "mavros/global_position/rel_alt", Float64, self.rel_alt_callback
+        )
+        self.compass_subscriber = rospy.Subscriber('mavros/global_position/compass_hdg', Float64, self.compass_callback)
+
+    def get_transform(self, base_frame="map", to_frame="camera_link"):
+        try:
+            transform = self.tf_buffer.lookup_transform(base_frame, to_frame, rospy.Time())
+            return transform
+        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException) as e:
+            rospy.logwarn(f"Failed to get transform: {e}")
+            return None
+
+    def bbox_to_ground_position(self, bbox: BoundingBoxLabeled):
+        ray_camera_frame = self.camera_model.projectPixelTo3dRay((bbox.bbox.center.x, bbox.bbox.center.y))
+        transform = self.get_transform("start_pose", "camera_link")
+        if transform:
+            vector = Vector3Stamped()
+            vector.vector.x = ray_camera_frame[0]
+            vector.vector.y = ray_camera_frame[1]
+            vector.vector.z = ray_camera_frame[2]
+            transformed_ray = do_transform_vector3(vector, transform)
+
+            camera_position = (transform.transform.translation.x, transform.transform.translation.y, transform.transform.translation.z)
+            ray_direction = (transformed_ray.vector.x, transformed_ray.vector.y, transformed_ray.vector.z)
+            ray_local_frame = pyrr.ray.create(camera_position, ray_direction)
+            plane = pyrr.plane.create()
+            intersection_point = pyrr.geometric_tests.ray_intersect_plane(ray_local_frame, plane)
+            if intersection_point is not None:
+                return intersection_point
+            rospy.logwarn("The line is parallel to the plane, cannot get intersection.")
+
+        return (0, 0, 0)
+
+    def create_figures(self, frame, bboxes: List[BoundingBoxLabeled], config):
+        def shrink_bbox(bbox_labeled: BoundingBoxLabeled, offset_percent: float = 0.0):
+            bbox_labeled.bbox.size_x = int(offset_percent * (bbox_labeled.bbox.size_x))
+            bbox_labeled.bbox.size_y = int(offset_percent * (bbox_labeled.bbox.size_y))
+
+        def get_img_piece(frame, bbox_labeled: BoundingBoxLabeled):
+            x1 = int(bbox_labeled.bbox.center.x - bbox_labeled.bbox.size_x / 2)
+            x2 = int(bbox_labeled.bbox.center.x + bbox_labeled.bbox.size_x / 2)
+            y1 = int(bbox_labeled.bbox.center.y - bbox_labeled.bbox.size_y / 2)
+            y2 = int(bbox_labeled.bbox.center.y + bbox_labeled.bbox.size_y / 2)
+            return frame[y1:y2, x1:x2]
```

Comment:
Move it outside the class

Reply:
I think it is fine, these functions are only used inside the parent function